lxd/container_console.go

```diff
@@ -0,0 +1,465 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"syscall"
+
+	"github.com/gorilla/mux"
+	"github.com/gorilla/websocket"
+	"gopkg.in/lxc/go-lxc.v2"
+
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/api"
+	"github.com/lxc/lxd/shared/logger"
+)
+
+type consoleWs struct {
+	// container currently worked on
+	container container
+
+	// uid to chown pty to
+	rootUid int64
+
+	// gid to chown pty to
+	rootGid int64
+
+	// websocket connections to bridge pty fds to
+	conns map[int]*websocket.Conn
+
+	// locks needed to access the "conns" member
+	connsLock sync.Mutex
+
+	// channel to wait until all websockets are properly connected
+	allConnected chan bool
+
+	// channel to wait until the control socket is connected
+	controlConnected chan bool
+
+	// map file descriptors -> secret
+	fds map[int]string
+
+	// terminal width
+	width int
+
+	// terminal height
+	height int
+}
+
+func (s *consoleWs) Metadata() interface{} {
+	fds := shared.Jmap{}
+	for fd, secret := range s.fds {
+		if fd == -1 {
+			fds["control"] = secret
+		} else {
+			fds[strconv.Itoa(fd)] = secret
+		}
+	}
+
+	return shared.Jmap{"fds": fds}
+}
+
+func (s *consoleWs) Connect(op *operation, r *http.Request, w http.ResponseWriter) error {
+	secret := r.FormValue("secret")
+	if secret == "" {
+		return fmt.Errorf("missing secret")
+	}
+
+	for fd, fdSecret := range s.fds {
+		if secret == fdSecret {
+			conn, err := shared.WebsocketUpgrader.Upgrade(w, r, nil)
+			if err != nil {
+				return err
+			}
+
+			s.connsLock.Lock()
+			s.conns[fd] = conn
+			s.connsLock.Unlock()
+
+			if fd == -1 {
+				s.controlConnected <- true
+				return nil
+			}
+
+			s.connsLock.Lock()
+			for i, c := range s.conns {
+				if i != -1 && c == nil {
+					s.connsLock.Unlock()
+					return nil
+				}
+			}
+			s.connsLock.Unlock()
+
+			s.allConnected <- true
+			return nil
+		}
+	}
+
+	/* If we didn't find the right secret, the user provided a bad one,
+	 * which 403, not 404, since this operation actually exists */
+	return os.ErrPermission
+}
+
+func (s *consoleWs) Do(op *operation) error {
+	<-s.allConnected
+
+	var err error
+	var ttys []*os.File
+	var ptys []*os.File
+
+	ttys = make([]*os.File, 1)
+	ptys = make([]*os.File, 1)
+	ptys[0], ttys[0], err = shared.OpenPty(s.rootUid, s.rootGid)
```

Comment:
you're never checking err

Reply:
I just realized that this check is also missing in `container_exec.go`.

Reply:
Pushed a fix for the same problem in `container_exec.go` on top of this branch.